### PR TITLE
Enable auto-reconnect

### PR DESCRIPTION
**Pull request by trestletech**
Silently tries to reconnect in the background for 3.5s, then shows the reconnecting screen with countdowns until the next reconnect attempt.

Out of scope for this PR:
 - UI improvements
 - "connect now" link

### CODE REVIEW

File: assets/shiny-server.js [hunk 1]
```diff
@@ -286,8 +284,8 @@
     // No an updated value like readyState, but rather a Boolean which will be set
     // true when the server has indicated that this connection can't ever be resumed.
     this._diconnected = false;
-    // True when the re/disconnect dialog is visible
-    this._disconnectDialog = false;
+    // The timer used to delay the display of the reconnecting dialog.
+    this._disconnectTimer = 0;
```
Comment (aronatkins) [inline assets/shiny-server.js]: Shouldn't this default to `null` or undefined?

File: assets/shiny-server.js [hunk 2]
```diff
@@ -305,8 +303,10 @@
     // be opened yet.
     this.onConnOpen = function() {
       self._first = false;
+      if (self._disconnectTimer){
+        clearTimeout(self._disconnectTimer);
```
Comment (aronatkins) [inline assets/shiny-server.js]: Reset `_disconnectTimer` back to its default value.

File: assets/shiny-server.js [hunk 3]
```diff
@@ -325,70 +325,149 @@
...
+          setReconnectDialog('Disconnected from server. Going to reconnect in ' + Math.ceil((targetTime-Date.now())/1000) + 's');
+        }
+        var msgTimer = setInterval(updateMsg, 500);
+        updateMsg(); // Run immediately
+        debug('Setting countdown timer');
+        debug('Scheduling reconnect attempt for ' + delay + 'ms');
+        // Schedule the reconnect for some time in the future.
+        setTimeout(function(){
+          var startTime = Date.now();
+          reconnect_p()
```
Comment (aronatkins) [inline assets/shiny-server.js]: nit: when reconnecting, maybe you should update the reconnect dialog saying that the attempt is happening.
```
setReconnectDialog('Attempting to reconnect ...')
```
Maybe show the number of prior attempts if `count > 1`

File: assets/shiny-server.js [hunk 4]
```diff
@@ -325,70 +325,149 @@
...
+          .then(function(){
+            // Able to reconnect.
+            return def;
```
Comment (aronatkins) [inline assets/shiny-server.js]: You need to clear `msgTimer` on success.

File: assets/shiny-server.js [hunk 5]
```diff
@@ -325,70 +325,149 @@
...
+          }, function(){
+            scheduleReconnect_p(startTime, count+1, expires, msgTimer)
+            .then(function(c){
+              def.resolve(c);
+            }, function(e){
+              def.reject(e);
+            })
+            .always(function(){
+              clearTimeout(msgTimer);
```
Comment (aronatkins) [inline assets/shiny-server.js]: nit: If you `clearTimeout` as the first thing each time through `scheduleReconnect_p`, you don't need the catch-all here. The current location is too late.

File: assets/shiny-server.js (hunk 4 above)
Comment (aronatkins) [inline assets/shiny-server.js]: Why the mix of `return def` and `def.resolve()`?

File: assets/shiny-server.js (hunk 3 above)
Comment (aronatkins) [inline assets/shiny-server.js]: Never mind. I found this in `reconnect_p`. It's a little hard to follow responsibilities/control. Sorry.

File: assets/shiny-server.js [hunk 6]
```diff
@@ -325,70 +325,149 @@
...
+        }
+
+        if (timer){
+          clearTimeout(timer);
         }
+        var targetTime = Date.now() + delay;
+        function updateMsg() {
+          setReconnectDialog('Disconnected from server. Going to reconnect in ' + Math.ceil((targetTime-Date.now())/1000) + 's');
+        }
+        var msgTimer = setInterval(updateMsg, 500);
```
Comment (aronatkins) [inline assets/shiny-server.js]: A followup to my other message about this `msgTimer`. It really should be canceled immediately before you call `reconnect_p`. That way, you don't need to worry about passing it through the recursion. In other words: Tick until you are about to reconnect and then cancel the ticker. Don't wait for reconnect success/failure.

This is more correct because your current timer may trigger while the reconnect is happening and incorrect update the reconnect dialog.

File: assets/shiny-server.js [hunk 7]
```diff
@@ -325,70 +325,149 @@
...
+        dialogMsg = msg;
+        $('#ss-connect-dialog').html(msg);
+      }
+
+      // Schedule the display of the disconnect window
+      self._disconnectTimer = setTimeout(function(){
+        debug('Displaying disconnect screen.');
+        $('body').addClass('ss-reconnecting');
+        $('<div id="ss-connect-dialog">'+dialogMsg+'<div class="ss-clearfix"></div></div><div id="ss-gray-out"></div>').appendTo('body');
+      }, 3500);
```
Comment (aronatkins) [inline assets/shiny-server.js]: Why `3500`?

Comment (aronatkins) [inline assets/shiny-server.js]: Sorry. I see your "3.5s silent reconnect" now.

Comment (aronatkins) [inline assets/shiny-server.js]: I'm a little uneasy about not showing the user any indication that reconnecting happened. Isn't it possible that messages in one direction or the other were lost? I'm aware that we do buffering, but does that cover in-flight messages at the time of the error?

I wonder if the right thing would be to show a reconnecting dialog always and then, on successful reconnect, show "Reconnected to your session" or something for a couple seconds.. which then fades away.

**Comment by aronatkins** [comment]
/review - just code at this point, haven't tested yet.

**Comment by trestletech** [comment]
Good call on the timer. This is much cleaner.

I hadn't thought about the in-flight message being dropped. That seems like such a rare case that I'd hate to disrupt the user experience to accommodate it, though.

How would you feel about a little spinner (but no gray overlay) that pops up center-screen when reconnecting, and perhaps some 2 second flash of a "successfully reconnected" whenever we reconnect.

(Bearing in mind that we don't really want to pull in any additional libraries here, so whatever we're doing need to be bare CSS + jQuery, so hopefully nothing too slick).

**Comment by aronatkins** [comment]
:+1: on the spinner + flash on reconnect. agreed about no libraries.

**Comment by aronatkins** [comment]
In practice, I'm not seeing this retry for more than about 10-13 seconds. You may need to let one more interval run. It's not a big deal if it's not exactly 15 seconds.

Timestamp for initial disconnect and final reconnect attempt:

Fri Dec 18 2015 13:38:36 GMT-0500 (EST)
Fri Dec 18 2015 13:38:49 GMT-0500 (EST)

**Comment by aronatkins** [comment]
Initial testing looks good. Let me know when you make a UI-improving pass.

**Comment by trestletech** [comment]
Yup. I'd expect 13 seconds on a fast connection. If we detect that our next poll is going to be after session expiration, then we schedule it instead for 2 seconds before session expiration, which I'm hoping is enough time to get the SockJS connection established before the session gets wiped from the server. After that, we'll only try once more as our final attempt, then we deem the connection closed since we're not going to try reconnecting again -- technically there's still 1.x seconds left on the session on a fast connection.

**Comment by trestletech** [comment]
After discussion with Winston this afternoon, I think it's actually better for us not to try to communicate anything about a silent reconnect. We'd need to communicate so much information to help them understand that a reconnect may be the source of some quirkiness (i.e. if they hit that brief window while a connection was dying but not registered as dead on the client yet). I think our options are to either do nothing or have a paragraph of text explaining what a reconnect is and why it's possible that they might observe some weirdness with their most recent change, blah blah blah. 

I think it's just better to go silent. So I'll merge w/o any new UI treatments, but we can revisit that decision.
